### PR TITLE
(PC-33222) feat(categories): handle native category positions

### PR DESCRIPTION
**Pull request by tconte-pass**
Link to JIRA ticket: https://passculture.atlassian.net/browse/PC-33222

## Flakiness

If I had to re-run tests in the CI due to flakiness, I add the incident [on Notion][2]

## Checklist

I have:

- [ ] Made sure my feature is working on web.
- [ ] Made sure my feature is working on mobile (depending on relevance : real or virtual devices)
- [ ] Written **unit tests** native (and web when implementation is different) for my feature.
- [ ] Added a **screenshot** for UI tickets or deleted the screenshot section if no UI change
- [ ] If my PR is a bugfix, I add the link of the "résolution de problème sur le bug" [on Notion][1]
- [ ] I am aware of all the best practices and respected them.

## Screenshots

**delete** _if no UI change_

| Platform         | Mockup/Before | After |
| :--------------- | :-----------: | :---: |
| iOS              |               |       |
| Android          |               |       |
| Phone - Chrome   | image  image | image image
            |
| Desktop - Chrome |               |       |

[1]: https://www.notion.so/passcultureapp/R-solution-de-probl-mes-sur-les-bugs-5dd6df8f6a754e6887066cf613467d0a
[2]: https://www.notion.so/passcultureapp/cb45383351b44723a6f2d9e1481ad6bb?v=10fe47258701423985aa7d25bb04cfee&pvs=4

## Best Practices

<details>
  <summary>Click to expand</summary>
These rules apply to files that you make changes to.
If you can't respect one of these rules, be sure to explain why with a comment.
If you consider correcting the issue is too time consuming/complex: create a ticket. Link the ticket in the code.

- In the production code: remove type assertions with `as` (type assertions are removed at compile-time, there is no runtime checking associated with a type assertion. There won’t be an exception or `null` generated if the type assertion is wrong). In certain cases `as const` is acceptable (for example when defining readonly arrays/objects). Using `as` in tests is tolerable.
- Remove bypass type checking with `any` (when you want to accept anything because you will be blindly passing it through without interacting with it, you can use `unknown`). Using `any` in tests is tolerable.
- Remove non-null assertion operators (just like other type assertions, this doesn’t change the runtime behavior of your code, so it’s important to only use `!` when you know that the value can’t be `null` or `undefined`).
- Remove all `@ts-expect-error` and `@eslint-disable`.
- Remove all warnings, and errors that we are used to ignore (`yarn test:lint`, `yarn test:types`, `yarn start:web`...).
- Use `gap` (`ViewGap`) instead of `<Spacer.Column />`, `<Spacer.Row />` or `<Spacer.Flex />`.
- Don't add new "alias hooks" (hooks created to group other hooks together). When adding new logic, this hook will progressively become more complex and harder to maintain.
- Remove logic from components that should be dumb.

Test specific:

- Avoid mocking internal parts of our code. Ideally, mock only external calls.
- When you see a local variable that is over-written in every test, mock it.
- Prefer `user` to `fireEvent`.
- When mocking feature flags, use `setFeatureFlags`. If not possible, mention which one(s) you want to mock in a comment (example: `jest.spyOn(useFeatureFlagAPI, 'useFeatureFlag').mockReturnValue(true) // WIP_NEW_OFFER_TILE in renderPassPlaylist.tsx` )
- In component tests, replace `await act(async () => {})` and `await waitFor(/* ... */)` by `await screen.findBySomething()`.
- In hooks tests, use `act` by default and `waitFor` as a last resort.
- Make a snapshot test for pages and modals ONLY.
- Make a web specific snapshot when your web page/modal is specific to the web.
- Make an a11y test for web pages.

Advice:

- Use TDD
- Use Storybook
- Use pair programming/mobs
</details>

### CODE REVIEW

File: src/features/search/pages/modals/CategoriesModal/CategoriesModal.tsx [hunk 1]
```diff
@@ -56,7 +57,7 @@ export const CategoriesModal = ({
   onClose,
   facets,
 }: CategoriesModalProps) => {
-  const { data } = useSubcategories()
+  const { data = PLACEHOLDER_DATA } = useSubcategories()
```
Comment (imouandjolobe-pass) [inline src/features/search/pages/modals/CategoriesModal/CategoriesModal.tsx:60]: Je ne pense pas que l'on veuille garder ce fallback, on se base sur des données écrites en dures, ce qui ne semble pas hyper fiable, alors qu'on voudrait uniquement se baser sur ce que nous retourne l'api
Je ne sais pas exactement quel changement dans la PR a rendu data potentiellement undefined par contre donc je n'ai pas de proposition à faire :/

File: src/features/search/helpers/categoriesHelpers/mapping-tree.ts [hunk 2]
```diff
@@ -176,7 +176,7 @@ export function createMappingTree(data: SubcategoriesResponseModelv2, facetsData
         : undefined
 
       result[searchGroup.name] = {
-        label: searchGroup.value || 'Toutes les catégories',
+        label: searchGroup.value ?? 'Toutes les catégories',
```
Comment (imouandjolobe-pass) [inline src/features/search/helpers/categoriesHelpers/mapping-tree.ts:159]: Je pense que l'on préfère avoir `||` ici, car si `searchGroup.value` vaut `''`, on va garder la string vide alors qu'avec `||`, on remplace par `Toutes les catégories`

**Review by imouandjolobe-pass — COMMENTED** [review]
Par contre, comment tu fais pour afficher le nouvel ordre des catégories ? J'ai lancé ta branche en local et je ne vois que le mapping actuel,, même en passant ta props `sort`à true

**Comment by tconte-pass** [comment]
> Par contre, comment tu fais pour afficher le nouvel ordre des catégories ? J'ai lancé ta branche en local et je ne vois que le mapping actuel,, même en passant ta props `sort`à true

En effet c'est pas clair, c'est parce que la PR côté backend qui renvoie des positions n'est pas encore mergée.
Elle est là : https://github.com/pass-culture/pass-culture-main/pull/15445
Faut tester en local pour l'instant pour le voir.

File: src/features/search/pages/modals/CategoriesModal/CategoriesModal.tsx (hunk 1 above)
Comment (tconte-pass) [reply src/features/search/pages/modals/CategoriesModal/CategoriesModal.tsx:60]: Je n'ai rien changé qui rende `data` possiblement `undefined`, ça l'était déjà. Et justement, le but de ce fallback c'est que ça ne puisse pas être le cas.
Même si c'est pas hyper fiable, c'est déjà ce qu'on met comme fallback dans `useSubcategories()` et on essaie quand même de tenir ce placeholder à jour.
C'est plus du typage, dans les faits la seule façon aujourd'hui que ce soit `undefined`, c'est en gros que le backend soit down.

Comment (imouandjolobe-pass) [reply src/features/search/pages/modals/CategoriesModal/CategoriesModal.tsx:60]: Il y a forcément eu un changement puisque depuis master, si je retire la ligne que tu as ajouté donc le = PLACEHOLDER_DATA, je n'ai pas d'erreur de typage dans le reste du fichier alors que sur ta branche, si 🤔
Et donc, si on met déjà ce PLACEHOLDER_DATA dans useSubcategories, je comprends encore moins pourquoi il faut le refaire ici

Comment (tconte-pass) [reply src/features/search/pages/modals/CategoriesModal/CategoriesModal.tsx:60]: C'est parce que j'ai modifié le typage de `createMappingTree` pour lequel `data` n'est plus optionnel.
Le but c'était de simplifier la gestion des `undefined`.
Parce qu'il y aucun cas dans lequel on peut travailler sans que `data` (et `tree`) ait une valeur.

Comment (bebstein-pass) [reply src/features/search/pages/modals/CategoriesModal/CategoriesModal.tsx:60]: le hook `useSubcategories` appel `useQuery`

d'après le type de retour, ça peut retourner `data: undefined` https://github.com/TanStack/query/blob/feb1dc91eace73b39fb6e34fac2ea505a9a57070/src/core/types.ts#L463 meme si on fourni un `placeholderData`

ce qui me semble faux, j'ai failli proposer de forcer le typage de retour de `useSubcategories`

par contre, en cas d'erreur, ça peut effectivement retourner un `data: undefined` https://github.com/TanStack/query/blob/feb1dc91eace73b39fb6e34fac2ea505a9a57070/src/core/types.ts#L478

donc on ne peut pas forcer le typage

cette proposition me convient

---

proposition alternative, pourrait-on facilement passer les subCategories en props du composant ?

**Review by bebstein-pass — APPROVED** [review]

File: src/features/search/helpers/categoriesHelpers/mapping-tree.ts [hunk 3]
```diff
@@ -79,42 +76,30 @@ export function getBooksGenreTypes(data: SubcategoriesResponseModelv2): OfferGen
...
+  return bookTree.reduce<MappedNativeCategories>((genresMapping, bookGenre) => {
+    const genreKey = getKeyFromStringLabel(bookGenre.label)
+    if (!genreKey) return genresMapping
+
+    genresMapping[genreKey] = {
+      ...bookGenre,
+      genreTypeKey: GenreType.BOOK,
+      children: bookGenre.children.reduce<MappedGenreTypes>((childrenMapping, child) => {
+        const childKey = getKeyFromStringLabel(child.label)
+        if (childKey) childrenMapping[childKey] = { ...child }
```
[the remark this answers is not in the document]
Comment (tconte-pass) [reply src/features/search/helpers/categoriesHelpers/mapping-tree.ts:94]: Je sais pas trop, mais le type de `child` est `BookSubType` et le type de `childrenMapping` est `MappedGenreTypes` a.k.a. `Record<string, MappedGenreType>`.
Les attributs de `BookSubType` sont inclus dans ceux de `MappedGenreType` :
```typescript
interface BookSubType {
  gtls: GTL[]
  label: string
  position: number
}
type MappedGenreType = {
  gtls?: GTL[]
  label: string
  position?: number
  nbResultsFacet?: never
}
```

J'adopte ta syntaxe sans problème si tu penses que c'est ok.

[the remark this answers is not in the document]
Comment (bebstein-pass) [reply src/features/search/helpers/categoriesHelpers/mapping-tree.ts:94]: si TS rale pas, c'est ok

si TS rale, c'est bizarre, parce qu'on ne fait que déconstruire et reconstruire sans changer le contenu

[the remark this answers is not in the document]
Comment (tconte-pass) [reply src/features/search/helpers/categoriesHelpers/mapping-tree.ts:94]: TS rale pas

**Pull request merged**